File: runtimes/python3/core/types/global_error.py
from typing import Optional, Dict, Any

class GlobalError(Exception):
    def __init__(self, msg: Optional[str] = None):
        Exception.__init__(self)
        # Convert exception objects to string for JSON serialization
        msg_str = str(msg) if msg is not None else None
        self.context: Dict[str, Any] = {'message': msg_str}
        self.message: str = msg_str
        self.code = 500

    def setCode(self, code: Optional[int] = None) -> None:
        self.code = code
        self.context['code'] = code

    def setJson(self, json: Optional[Dict[str, Any]] = None) -> None:
        self.context['json'] = json

    def setStack(self, stack: Optional[str] = None) -> None:
        self.context['stack'] = stack

    def setName(self, name: Optional[str] = None) -> None:
        self.context['name'] = name

    def hasJson(self) -> bool:
        return 'json' in self.context

    def __str__(self) -> str:
        if 'json' in self.context:
            return str(self.context['json'])
        return self.context['message'] or ''
    
    def to_dict(self):
        return self.context

File: core/shared/src/GlobalError.ts
import type ErrorContext from "./types/ErrorContext";
import type ParamsDictionary from "./types/ParamsDictionary";

export default class GlobalError extends Error {
	public context: ErrorContext = { message: "" };

	constructor(msg: string | undefined) {
		super(msg);
		// Standard Error-subclass pattern: pin to new.target's prototype (the
		// class actually being constructed), not GlobalError's own — otherwise
		// this clobbers a subclass's correct prototype and `instanceof Subclass`
		// silently fails for any subclass that doesn't re-pin itself (#736).
		Object.setPrototypeOf(this, new.target.prototype);

		this.context.message = msg as string;
	}

	setCode(code?: number) {
		this.context.code = code;
	}
	setJson(json?: Record<string, unknown>) {
		this.context.json = json as ParamsDictionary;
	}
	setStack(stack?: string) {
		this.context.stack = stack;
	}
	setName(name?: string) {
		this.context.name = name;
	}
	/** #996 — response headers to emit with this error (e.g. `Location`). */
	setHeaders(headers?: Record<string, string>) {
		this.context.headers = headers;
	}
	/** #996 — raw `Set-Cookie` values to emit with this error. */
	setCookies(cookies?: string[]) {
		this.context.cookies = cookies;
	}

	hasJson(): boolean {
		return this.context.json !== undefined;
	}

	override toString(): string {
		if (this.context.json) return JSON.stringify(this.context.json);
		return this.context.message as string;
	}
}
